internal/common/common.go

```diff
@@ -144,11 +150,243 @@ func GetVaultAuthNamespaced(ctx context.Context, c client.Client, obj client.Obj
+				return nil, nil, err
+			}
+			globalAuthMount = globalAuthMethod.Mount
+			globalAuthNamespace = globalAuthMethod.Namespace
+			globalAuthParams = globalAuthMethod.Params
+			globalAuthHeaders = globalAuthMethod.Headers
+		}
+	case vaultcredsconsts.ProviderMethodAWS:
+		globalAuthMethod := gObj.Spec.AWS
+		mergeTargetAuthMethod := cObj.Spec.AWS
+		if mergeTargetAuthMethod == nil && globalAuthMethod == nil {
+			return nil, nil, fmt.Errorf("global auth method %s is not configured "+
+				"in VaultAuthGlobal %s", cObj.Spec.Method, authGlobalRef)
+		}
+
+		if globalAuthMethod != nil {
+			srcAuthMethod := globalAuthMethod.VaultAuthConfigAWS.DeepCopy()
+			if mergeTargetAuthMethod == nil {
+				cObj.Spec.AWS = srcAuthMethod
+			} else {
+				merged, err := mergeTargetAuthMethod.Merge(srcAuthMethod)
+				if err != nil {
+					return nil, nil, err
+				}
+				cObj.Spec.AWS = merged
+			}
+			if err := cObj.Spec.AWS.Validate(); err != nil {
+				return nil, nil, err
+			}
+			globalAuthMount = globalAuthMethod.Mount
+			globalAuthNamespace = globalAuthMethod.Namespace
+			globalAuthParams = globalAuthMethod.Params
+			globalAuthHeaders = globalAuthMethod.Headers
+		}
+	case vaultcredsconsts.ProviderMethodGCP:
+		globalAuthMethod := gObj.Spec.GCP
+		mergeTargetAuthMethod := cObj.Spec.GCP
+		if mergeTargetAuthMethod == nil && globalAuthMethod == nil {
+			return nil, nil, fmt.Errorf("global auth method %s is not configured "+
+				"in VaultAuthGlobal %s", cObj.Spec.Method, authGlobalRef)
+		}
+
+		if globalAuthMethod != nil {
+			srcAuthMethod := globalAuthMethod.VaultAuthConfigGCP.DeepCopy()
+			if mergeTargetAuthMethod == nil {
+				cObj.Spec.GCP = srcAuthMethod
+			} else {
+				merged, err := mergeTargetAuthMethod.Merge(srcAuthMethod)
+				if err != nil {
+					return nil, nil, err
+				}
+				cObj.Spec.GCP = merged
+			}
+			if err := cObj.Spec.GCP.Validate(); err != nil {
+				return nil, nil, err
+			}
+			globalAuthMount = globalAuthMethod.Mount
+			globalAuthNamespace = globalAuthMethod.Namespace
+			globalAuthParams = globalAuthMethod.Params
+			globalAuthHeaders = globalAuthMethod.Headers
+		}
+	default:
+		return nil, nil, fmt.Errorf(
+			"unsupported auth method %q for global auth merge",
+			cObj.Spec.Method,
+		)
+	}
+
+	cObj.Spec.Mount = firstNonZeroLen(strLenFunc,
+		cObj.Spec.Mount, globalAuthMount, gObj.Spec.DefaultMount)
+	if cObj.Spec.Mount == "" {
+		return nil, nil, fmt.Errorf(
+			"mount is not set in VaultAuth %s after merge with %s",
+			client.ObjectKeyFromObject(cObj), authGlobalRef,
+		)
+	}
+
+	cObj.Spec.Namespace = firstNonZeroLen(strLenFunc,
+		cObj.Spec.Namespace, globalAuthNamespace, gObj.Spec.DefaultVaultNamespace)
+	cObj.Spec.Headers = firstNonZeroLen(mapLenFunc[string, string],
+		cObj.Spec.Headers, globalAuthHeaders, gObj.Spec.DefaultHeaders)
+	cObj.Spec.Params = firstNonZeroLen(mapLenFunc[string, string],
+		cObj.Spec.Params, globalAuthParams, gObj.Spec.DefaultParams)
```

Comment:
It was definitely a consideration to take the union of all params/headers, but I hesitant, since there would be no way for a referring VaultAuth to omit specific keys without extending the functionality of the vaultAuthGlobalRef. One thing that we could do is to allow for configurable merge strategies to be set in the VaultAuthSpec. I can put up a spike PR for that.

Reply:
@tvoran put up a spike PR here: https://github.com/hashicorp/vault-secrets-operator/pull/757